File: scripts/inspect_payload.py
"""
scripts/inspect_payload.py

Python helper to inspect migration-payload.json and extract hub-like items.

Usage:
  python3 scripts/inspect_payload.py migration-payload.json
  python3 scripts/inspect_payload.py migration-payload.json --extract-hubs
"""

import json
import sys
import os
from typing import Any, Dict, List

def read_json(fp: str) -> Any:
    with open(fp, 'r', encoding='utf-8') as f:
        return json.load(f)

H_KEYS = ['hub','city','country','country_code','lat','lng','lon','latitude','longitude','location','coords','postal','zip','name']

def detect_fields(obj: Any, prefix='') -> List[str]:
    found = set()
    if isinstance(obj, dict):
        for k, v in obj.items():
            lk = k.lower()
            for h in H_KEYS:
                if h in lk:
                    found.add(prefix + k if prefix else k)
            # recurse
            if isinstance(v, (dict, list)):
                for sub in detect_fields(v, prefix + k + '.' if prefix else k + '.'):
                    found.add(sub)
            elif isinstance(v, str):
                vs = v.lower()
                for h in ['hub','city','country','latitude','longitude']:
                    if h in vs:
                        found.add(f'(value contains {h})')
    elif isinstance(obj, list):
        for i, el in enumerate(obj):
            for sub in detect_fields(el, f'{prefix}[{i}].'):
                found.add(sub)
    return list(found)

def extract_latlon(obj: Any):
    def to_num(v):
        if v is None: return None
        try:
            return float(str(v).replace(',', '.'))
        except:
            return None
    if isinstance(obj, dict):
        lat = lon = None
        for k, v in obj.items():
            lk = k.lower()
            if any(x in lk for x in ['lat','latitude']):
                lat = to_num(v) or lat
            if any(x in lk for x in ['lon','lng','longitude']):
                lon = to_num(v) or lon
            if isinstance(v, dict):
                nested = extract_latlon(v)
                if nested:
                    lat = lat or nested.get('lat')
                    lon = lon or nested.get('lon')
        if lat is not None or lon is not None:
            return {'lat': lat, 'lon': lon}
    return None

def normalize_payload(raw):
    if isinstance(raw, list):
        return raw
    if isinstance(raw, dict):
        if 'collections' in raw and isinstance(raw['collections'], dict):
            all_items = []
            for k, v in raw['collections'].items():
                if isinstance(v, list):
                    for it in v:
                        if 'collection_name' not in it:
                            it['collection_name'] = k
                        all_items.append(it)
            if all_items:
                return all_items
        # fallback: find first array of objects
        for k, v in raw.items():
            if isinstance(v, list) and v and isinstance(v[0], dict):
                return v
    return []

def main():
    if len(sys.argv) < 2 or '--help' in sys.argv or '-h' in sys.argv:
        print('Usage: python3 scripts/inspect_payload.py migration-payload.json [--extract-hubs]')
        sys.exit(0)
    fp = sys.argv[1]
    extract = '--extract-hubs' in sys.argv
    if not os.path.exists(fp):
        print('File not found:', fp)
        sys.exit(2)
    raw = read_json(fp)
    items = normalize_payload(raw)
    print('Loaded payload. Item count:', len(items))
    # collections summary
    colcounts = {}
    for it in items:
        cname = it.get('collection_name') if isinstance(it, dict) else 'unknown'
        colcounts[cname] = colcounts.get(cname, 0) + 1
    print('\\nCollections summary:')
    for k, v in sorted(colcounts.items(), key=lambda kv: -kv[1]):
        print(f' - {k}: {v}')
    print('\\nSample previews:')
    for i, it in enumerate(items[:20]):
        preview = json.dumps(it.get('item', it))[:200]
        print(f'{i+1}. collection: {it.get("collection_name","unknown")} preview: {preview}')
    # detect hub-like
    hubs = []
    for it in items:
        obj = it.get('item', it) if isinstance(it, dict) else it
        fields = detect_fields(obj)
        latlon = extract_latlon(obj)
        if fields or latlon:
            hubs.append({'collection_name': it.get('collection_name','unknown'), 'id': it.get('id'), 'fields': fields, 'latlon': latlon, 'item': obj})
    print('\\nFound', len(hubs), 'hub-like candidate(s).')
    for i, h in enumerate(hubs[:50]):
        print(f'\\n[{i+1}] collection: {h["collection_name"]} id:{h.get("id","-")} fields:{",".join(h["fields"])} latlon:{h.get("latlon")}')
        print(' preview:', json.dumps(h['item'])[:400])
    if extract:
        out = 'hubs_candidates.json'
        with open(out, 'w', encoding='utf-8') as f:
            json.dump(hubs, f, indent=2, ensure_ascii=False)
        print('\\nWrote hubs candidates to', out)
    else:
        print('\\nTip: re-run with --extract-hubs to write hubs_candidates.json') 

if __name__ == '__main__':
    main()

File: scripts/inspectPayload.js
/**
 * scripts/inspectPayload.js
 *
 * Node.js helper to inspect migration-payload.json and extract "hub-like" items.
 *
 * Usage:
 *   node scripts/inspectPayload.js migration-payload.json
 *   node scripts/inspectPayload.js migration-payload.json --extract-hubs
 *
 * Responsibilities:
 * - Load the payload (array or object with collections).
 * - Print collection names and counts.
 * - Search for hub-like items (city/country/lat/lon/hub/location).
 * - Optionally write hubs_candidates.json with matching items.
 */

/**
 * @description Synchronously read JSON from a file path
 * @param {string} fp - file path
 * @returns {any} parsed JSON
 */
function readJson(fp) {
  const fs = require('fs');
  const path = require('path');
  const resolved = path.resolve(fp);
  if (!fs.existsSync(resolved)) {
    throw new Error(`File not found: ${resolved}`);
  }
  const raw = fs.readFileSync(resolved, 'utf8');
  return JSON.parse(raw);
}

/**
 * @description Heuristic: detect hub-like keys in an object (case-insensitive)
 * @param {any} obj - object or primitive
 * @returns {string[]} list of matched keys or [] if none
 */
function detectHubLikeFields(obj) {
  const found = new Set();
  const heuristics = [
    'hub',
    'city',
    'country',
    'country_code',
    'countrycode',
    'countrycode',
    'lat',
    'lng',
    'lon',
    'latitude',
    'longitude',
    'location',
    'coords',
    'postal',
    'zip',
    'name'
  ];

  function scan(o, prefix = '') {
    if (o == null) return;
    if (typeof o === 'object') {
      for (const k of Object.keys(o)) {
        const lk = k.toLowerCase();
        for (const h of heuristics) {
          if (lk.includes(h)) {
            found.add(prefix ? `${prefix}.${k}` : k);
          }
        }
        // Recurse if nested
        try {
          scan(o[k], prefix ? `${prefix}.${k}` : k);
        } catch (e) {
          // ignore
        }
      }
    } else if (typeof o === 'string') {
      const s = o.toLowerCase();
      for (const h of ['hub', 'city', 'country', 'latitude', 'longitude']) {
        if (s.includes(h)) {
          found.add(`(value contains "${h}")`);
        }
      }
    }
  }

  scan(obj);
  return Array.from(found);
}

/**
 * @description Try to coerce lat/lon numeric pairs from an object
 * @param {any} obj
 * @returns {{lat?:number, lon?:number}|null}
 */
function extractLatLon(obj) {
  if (obj == null || typeof obj !== 'object') return null;
  const candidates = {};

  const tryNum = (v) => {
    if (v == null) return null;
    const n = Number(String(v).replace(',', '.'));
    return Number.isFinite(n) ? n : null;
  };

  // Common key names
  const latKeys = ['lat', 'latitude'];
  const lonKeys = ['lon', 'lng', 'longitude'];

  for (const lk of latKeys) {
    for (const k of Object.keys(obj)) {
      if (k.toLowerCase().includes(lk)) {
        const n = tryNum(obj[k]);
        if (n != null) candidates.lat = n;
      }
    }
  }

  for (const lk of lonKeys) {
    for (const k of Object.keys(obj)) {
      if (k.toLowerCase().includes(lk)) {
        const n = tryNum(obj[k]);
        if (n != null) candidates.lon = n;
      }
    }
  }

  // If not at top level, try nested objects
  if ((candidates.lat == null || candidates.lon == null)) {
    for (const k of Object.keys(obj)) {
      const v = obj[k];
      if (v && typeof v === 'object') {
        const nested = extractLatLon(v);
        if (nested) {
          if (nested.lat != null) candidates.lat = nested.lat;
          if (nested.lon != null) candidates.lon = nested.lon;
        }
      }
    }
  }

  if (candidates.lat != null || candidates.lon != null) {
    return { lat: candidates.lat, lon: candidates.lon };
  }
  return null;
}

/**
 * @description Flatten the payload into an array of migration items
 * @param {any} raw
 * @returns {Array<{collection_name?:string, item?:any, id?:string}>}
 */
function normalizePayload(raw) {
  // If it's already an array of items (common case)
  if (Array.isArray(raw)) return raw;

  // If it's an object with collections
  if (raw && typeof raw === 'object') {
    // Common shapes: { collections: { migration_items: [...] } }
    if (raw.collections && typeof raw.collections === 'object') {
      const all = [];
      for (const k of Object.keys(raw.collections)) {
        const v = raw.collections[k];
        if (Array.isArray(v)) {
          v.forEach((it) => {
            // tag collection name if missing
            if (!it.collection_name) it.collection_name = k;
            all.push(it);
          });
        }
      }
      if (all.length > 0) return all;
    }

    // Try to find first array of objects that looks like migration items
    for (const k of Object.keys(raw)) {
      if (Array.isArray(raw[k]) && raw[k].length > 0 && typeof raw[k][0] === 'object') {
        return raw[k];
      }
    }
  }

  return [];
}

/**
 * @description Main entry
 */
function main() {
  const fs = require('fs');
  const path = require('path');

  const argv = process.argv.slice(2);
  if (argv.length === 0 || argv.includes('--help') || argv.includes('-h')) {
    console.log('Usage: node scripts/inspectPayload.js <migration-payload.json> [--extract-hubs]');
    process.exit(0);
  }

  const file = argv[0];
  const extractHubs = argv.includes('--extract-hubs');

  let raw;
  try {
    raw = readJson(file);
  } catch (err) {
    console.error('Failed to read JSON:', err.message);
    process.exit(2);
  }

  const items = normalizePayload(raw);
  console.log(`Loaded payload. Item count: ${items.length}`);

  // List collection_name counts
  const collections = new Map();
  for (const it of items) {
    const cname = it.collection_name || (it.collection ? it.collection : 'unknown');
    collections.set(cname, (collections.get(cname) || 0) + 1);
  }

  console.log('\\nCollections summary:');
  for (const [cname, cnt] of Array.from(collections.entries()).sort((a, b) => b[1] - a[1])) {
    console.log(`  - ${cname}: ${cnt}`);
  }

  // Quick sample keys (first 20)
  console.log('\\nSample items (first 20 previews):');
  items.slice(0, 20).forEach((it, idx) => {
    const cname = it.collection_name || 'unknown';
    const short = JSON.stringify(it.item ?? it).slice(0, 200).replace(/\\n/g, '');
    console.log(`${idx + 1}. collection: ${cname} — preview: ${short}`);
  });

  // Hub detection
  const hubCandidates = [];
  for (const it of items) {
    const candidateObject = it.item || it; // sometimes item is nested, sometimes the whole object
    const fields = detectHubLikeFields(candidateObject);
    const latlon = extractLatLon(candidateObject);
    const signature = fields.length > 0 || latlon != null;

    if (signature) {
      hubCandidates.push({
        collection_name: it.collection_name || 'unknown',
        id: it.id || it.ID || null,
        fields,
        latlon,
        preview: JSON.stringify(candidateObject).slice(0, 800),
        raw: candidateObject
      });
    }
  }

  console.log(`\\nFound ${hubCandidates.length} hub-like candidate(s).`);
  if (hubCandidates.length > 0) {
    hubCandidates.slice(0, 50).forEach((c, i) => {
      console.log(`\\n[${i + 1}] collection: ${c.collection_name} id:${c.id || '-'} fields:${c.fields.join(', ')} latlon:${c.latlon ? JSON.stringify(c.latlon) : '-'}`);
      console.log(` preview: ${c.preview.slice(0, 400)}`);
    });

    if (extractHubs) {
      const outPath = path.resolve('hubs_candidates.json');
      fs.writeFileSync(outPath, JSON.stringify(hubCandidates.map(h => ({ collection_name: h.collection_name, id: h.id, fields: h.fields, latlon: h.latlon, item: h.raw })), null, 2), 'utf8');
      console.log(`\\nWrote ${hubCandidates.length} candidates to ${outPath}`);
    } else {
      console.log('\\nTip: re-run with --extract-hubs to write hubs_candidates.json for inspection.');
    }
  } else {
    console.log('\\nNo hub-like items detected. If you expected hubs, consider these checks:');
    console.log('- Are you running this script on the same migration-payload.json you downloaded?');
    console.log('- Did the app store hubs in IndexedDB instead of localStorage? Use the IndexedDB extractor if so.');
  }
}

main();